### ISSUE TITLE
availability-recovery: systemic chunks recovery hotpath

### ISSUE THREAD

**Issue by sandreim** [paritytech/polkadot#7505]
Currently we randomize the chunks that we fetch and we don't reap any benefits from our erasure coding being systemic (chunks `0..f+1` are the exact copy of the PoV).

We should add another hotpath for doing so. This would probably go to the decoder and make the `reconstruct` very cheap or otherwise not call it at all, and have the subsystem assemble the PoV. 

One concern here is that the we only change the validator indices on every session and these indices map directly to the systemic chunks indices right now. We need to spread the load across all validators in order to use the systemic code, so we'd have to change the way the validator indices who have the systemic chunks are mapped. We could make this mapping to depend on something other than just validator index which must be feasible to obtain in all contexts - approval checking and disputes.

CC @burdges for more suggestions on the new mapping

**Comment by burdges** [comment]
> we only change the validator indices on every session and these indices map directly to the systemic chunks indices right now.

We screwed this up then.  We should've permuted the map between validator indices and chunk indices based upon the epoch randomness and slot.  I'd asked for this originally, and proposed code, but I guess it got omitted.

**Comment by ordian** [comment]
> One concern here is that the we only change the validator indices on every session and these indices map directly to the systemic chunks indices right now

are you sure that's the case? I remember hearing different from @drahnr 
also you can test that here https://github.com/paritytech/polkadot/blob/f900dede87efcfd91eb5656ed78077ed3877f323/erasure-coding/src/lib.rs#L362

**Comment by sandreim** [comment]
Looked at the code and ValidatorIndex is used as chunk index.

**Comment by ordian** [comment]
Yes, but this is not true:
> chunks 0..f+1 are the exact copy of the PoV

(or that there's a subset of f+1 chunks that is the exact PoV clone)

**Comment by sandreim** [comment]
Ok, I see what you mean, but as per discussion with @burdges this was the design and should be true. My assumption was the implementation does that (didn't check)

**Comment by rphmeier** [comment]
The two breaking changes that can go in at the same time:
  1. Systemic chunks implementation
  2. Permuting validator indices in erasure-coding for each available datum (probably not the relay-parent, as with asynchronous backing + #7441 we may be bundling many different relay parents together on a single availability core)

FWIW when it comes to building fast-paths it'd also be good to have approval checkers help others get the systemic chunks.

**Comment by burdges** [comment]
> (or that there's a subset of f+1 chunks that is the exact PoV clone)

This should be true.

I wanted the relay chain slot and epoch randomness in this mapping anyways.  If we use the block height instead, then we can easily code it to activate at a specific block height, and likely we can remove the dependency upon the epoch randomness.
```
use fpe::ff1::{FF1, FlexibleNumeralString};

pub enum AvailabilityFF1(Option<FF1>);

impl AvailabilityFF1 {
    fn new(blockheight: u64, num_validators: u16) -> AvailabilityFF1 {
        if blockheight < WHATEVER { return AvailabilityFF1(None); }
        use rand::{SeedableRng, seq::SliceRandom};
        let mut seed = [0u8; 32];
        seed[0..4].copy_from_slice(& blockheight.to_by_bytes());
        let mut rng = rand_chacha::ChaChaRng::from_seed(seed);
        // We could create a map table like this
        // let mut map: Vec<u16> = (0..n).collect();
        // map.as_mut_slice().shuffle(&mut rng);
        // but in theory format preserving encryption might save allocations and cache,
        // although not with its current Vec usage in the fpe crate.
        rng.fill_bytes(&mut seed);
        let ff1 = FF1::<aes::Aes256>::new(&seed, num_validators).unwrap();
        AvailabilityFF1(Some(ff1))
    }

    // FlexibleNumeralString is ridiculously inefficient here, but we could
    // send the fpe crate a PR that impls their NumeralString and
    // Operations for arrayvec::ArrayVec;

    fn map_validator_to_chunk(&self, validator_index: u16) -> u16 {
        let ff1 = match self { Some(ff1) => ff1, None => return validator_index, }
        let pt: FlexibleNumeralString = vec![validator_index; 1];
        ff1.encrypt(&[], &::from_bytes_le(&pt)).unwrap()[0]
    }

    fn map_chunk_to_validator(&self, chunk_index: u16) -> u16 {
        let ff1 = match self { Some(ff1) => ff1, None => return chunk_index, }
        let ct: FlexibleNumeralString = vec![validator_index; 1];
        ff1.decrypt(&[], &::from_bytes_le(&ct)).unwrap()[0]
    }
}
```
We do not afaik care that adversaries can place their own validators in systemic positions for specific blocks.

**Comment by sandreim** [comment]
FWIW, we can still use the hot-path of fetching from backers at least to some higher upper limit than now to achieve the same cost reduction as systemic chunks.

**Comment by burdges** [comment]
Yeah, you can fetch from them even if they're not randomized, but it risks overloading them of course, hence the randomization.

**Comment by sandreim** [comment]
Currently, the backing group is shuffled before we attempt fetching the available data.

**Comment by burdges** [comment]
oops, sorry I miss-read your comment above.  Yeah I think you can fetch the systemic chunks by index too, if you try harder to get them from backers.

We can also just put a shuffle into the indices, like was originally supposed to be there.
